### ISSUE TITLE
Improve error spans for certain literals

### ISSUE THREAD

**Issue by RyanCavanaugh** [Microsoft/TypeScript#19579]
```
// Problem 1: Error spans are useless in contextually-typed array/object literal cases
type Kind = 'foo' | 'ball';
// Error is on declared identifier instead of bad element
const ka: Kind[] = [
    'foo', 'ball', 'xxxx', 'foo', 'ball'
];
// Problem 2: Error is on declared identifier instead of bad key
const ko: { [s: string]: Kind } = {
    'x': 'foo',
    'y': 'xxxx',
    'z': 'foo'
};
```

**Comment by mattmccutchen** [comment]
This has been bugging me for a while and I finally went to file a suggestion.  Fortunately, I found yours.

Another case:
```typescript
// Error is on `f`.
let f: (x: number) => number = (x) => {
  if (x % 2 == 0)
    return x + 1;
  else
    return "oops";  // Error should be marked here.
}
```

An extreme case is [a giant object literal that serves as a dispatch table for an interpreter](https://bitbucket.org/objsheets/objsheets/src/d45748954c829a1a1cc5ce9a2dcafc534b465c96/src/language/formulas.ts?at=master&fileviewer=file-view-default#formulas.ts-585).  I also have some [object literals that are passed directly to a function and not assigned to a variable](https://bitbucket.org/objsheets/objsheets/src/d45748954c829a1a1cc5ce9a2dcafc534b465c96/src/layout/binding.ts?at=master&fileviewer=file-view-default#binding.ts-307), and in those cases, TypeScript squiggles the whole object literal if there is any type error.  Aside from the difficulty in locating the type error, it's distracting to see the whole object literal squiggle and unsquiggle as I'm typing, and the squiggle on the whole object literal makes it harder to see squiggles for any other errors within the object literal.

I'll be happy to help with a proposal if there's something I can do that provides a reasonable return for time spent; you probably know much more about what is needed than I do.
